File: code/ex15.py
class Solution:
    def threeSum(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """

        nums.sort()
        length = len(nums)
        List = []

        for i in range(0, length - 2):
            for j in range(i + 1, length - 1):
                if ((0 - nums[i] - nums[j]) in nums[j + 1:length + 1:1]):
                    l = [nums[i], nums[j], 0 - nums[i] - nums[j]]
                    if l not in List:
                        List.append(l)
        
        return List
        
sol = Solution()
print(sol.threeSum([-1, 0, 1, 2, -1, -4]))

File: code/ex15.cpp
// Two pointers

class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        int n = nums.size();
        vector<vector<int>> res;
        
        for (int i = 0; i < n; i++) {
            if (i > 0 && nums[i - 1] == nums[i])    continue;    // Skip duplicates
            int lo = i + 1, hi = n - 1;
            int target = -nums[i];
            
            while (lo < hi) {    // Use two pointers to find the other two numbers
                if (nums[lo] + nums[hi] == target) {
                    res.push_back({nums[i], nums[lo], nums[hi]});
                    // Skip duplicates
                    while (lo < hi && nums[lo + 1] == nums[lo])    lo++;
                    while (lo < hi && nums[hi - 1] == nums[hi])    hi--;
                    lo++;    hi--;
                }
                else if (nums[lo] + nums[hi] < target) {
                    lo++;
                }
                else {
                    hi--;
                }
            }
        }
        
        return res;
    }
};